Market update 5

limichange opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

**New Features**
  * Introduced responsive transaction history layouts with normal and compact views, adapting to screen size and column width.
  * Added horizontal scrolling for transaction lists on large screens with limited width.
  * Implemented new components for transaction amounts and addresses, with interactive copy and external view actions.
  * Added support for capping market cap values at 999T in number formatting and display.
  * Updated default liquidity filter in Market Home to a minimum of 5K.
  * Added automatic formatter selection in number displays based on value thresholds.
  * Added state management for sorting in market token and watchlist hooks, enabling external control.
  * Enabled natural ordering of watchlist tokens by sort index when no sorting is applied.
  * Enhanced watchlist additions to assign sort index positioning new items at the top.
  * Introduced new toggle state and actions for showing watchlist-only mode with app-wide event emission.
  * Added new context providers for market watchlist state management with improved store handling.

**Improvements**
  * Enhanced market cap and price formatting with new capping and display logic.
  * Updated search query generation for token symbols to include a dollar sign.
  * Improved test coverage for number formatting with new cases for capped values.
  * Refactored and reorganized transaction history components for better maintainability and performance.
  * Enhanced market capitalization display in token details with refined formatting and component order.
  * Modified market token list component to delegate sorting and watchlist display control externally.
  * Added dynamic tracking of left column width for responsive layout adjustments.
  * Refactored watchlist add/remove actions for immediate local state updates and fire-and-forget API calls.
  * Simplified watchlist toggle controls by internalizing state management within toggle components.
  * Consolidated exports and reorganized modules for transaction history layouts and components.
  * Added locale and timezone awareness to TradingView component for improved regional display.
  * Centralized locale and timezone mapping logic for TradingView integration.

**Bug Fixes**
  * Fixed display of very large market cap values to prevent overflow by capping at 999T.

**Chores**
  * Added 'nodata' to the spell checker skip word list.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

limichange committed:
refactor: Update TransactionItem component layout for improved price display

- Replaced SizableText with NumberSizeableText for better price formatting in the TransactionItem component.
- Enhanced layout by introducing XStack for clearer separation of base and quote amounts, improving overall readability and user experience.

limichange committed:
feat: Enhance number formatting with capAtMaxT option

- Introduced capAtMaxT functionality in number formatting utilities to limit market cap display to a maximum of 999T.
- Updated formatMarketCap and formatDisplayNumber functions to support this new feature, including isCapped metadata.
- Enhanced tests to validate capAtMaxT behavior across various scenarios, ensuring accurate formatting and user feedback.

limichange committed:
feat: Update price formatting logic in MarketTokenList for better clarity

- Enhanced the price rendering logic in the useColumnsDesktop hook to utilize BigNumber for determining the appropriate formatter based on value thresholds, improving the display of market cap and price.
- This change ensures that larger values are formatted as market cap, enhancing user understanding of token values.

limichange committed:
feat: Refactor TokenDetailHeaderRight for improved market cap formatting

- Updated the TokenDetailHeaderRight component to utilize formatDisplayNumber and formatMarketCap for displaying market cap values, enhancing clarity and consistency in the UI.
- This change replaces the previous numberFormat usage, ensuring better formatting for market cap display.

limichange committed:
fix: Ensure correct type for market cap display in TokenDetailHeaderRight

- Updated the value assignment in the TokenDetailHeaderRight component to explicitly cast the result of formatDisplayNumber as a string, ensuring type consistency and preventing potential rendering issues.

limichange committed:
refactor: Simplify render logic in useColumnsDesktop for market token change display

- Updated the render function for the 'change24h' column in the useColumnsDesktop hook to use a more explicit return statement, enhancing code readability and maintainability.

limichange committed:
refactor: Remove unused number formatting utility from TransactionItem component

- Eliminated the import of numberFormat from the TransactionItem component, streamlining the code and improving maintainability.

revan-zhang commented:
![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/pr-banner-default.svg)

### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/rayston92/pr-checks/c38bfcc7-b9dc-4cbd-ba58-8a9e77349f3c?refs=b6a4cbba9e42d074dc796aa07b6b2591342f20ee&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/rayston92/pr-checks/c38bfcc7-b9dc-4cbd-ba58-8a9e77349f3c/license?refs=b6a4cbba9e42d074dc796aa07b6b2591342f20ee&source=prChecksComment)

limichange committed:
Merge branch 'x' into market-update-5

limichange committed:
feat: Add examples for marketCap formatting with capAtMaxT in NumberSizeableTextGallery

- Introduced new examples demonstrating the capAtMaxT functionality for market cap display in the NumberSizeableTextGallery component.
- Enhanced the component by adding visual comparisons for values with and without the cap, improving clarity for users regarding large number formatting.

limichange committed:
fix: Correct share button placement in TokenDetailHeaderRight component

- Adjusted the rendering order of the share button and market star in the TokenDetailHeaderRight component to ensure proper display and alignment in the UI.

limichange committed:
refactor: Improve layout and spacing in NumberSizeableTextGallery component

- Adjusted spacing and layout in the NumberSizeableTextGallery to enhance visual clarity and consistency.
- Updated gap values in YStack and XStack components for better alignment of text elements.

limichange committed:
refactor: Enhance TransactionsHistory layout with horizontal scrolling

- Wrapped the ListView in a ScrollView to enable horizontal scrolling for transaction history, improving user experience when viewing long lists of transactions.
- Updated imports to include ScrollView from components for this enhancement.

limichange committed:
refactor: Restructure TransactionsHistory component for improved modularity

- Removed deprecated components (TransactionItem, TransactionsHeader, TransactionsSkeleton) and replaced them with a modular structure using new components for better maintainability.
- Introduced layout variations (TransactionItemNormal and TransactionItemSmall) to enhance responsiveness based on screen size.
- Updated the TransactionsHistory component to utilize the new layout and component structure, improving code clarity and organization.

limichange committed:
refactor: Update TransactionsHistory layout for improved responsiveness

- Modified the TransactionsHistory component to adjust the ScrollView's content container style for better vertical alignment.
- Ensured the TransactionsHeaderNormal component spans the full width for enhanced layout consistency.
- Cleaned up imports in TransactionItemSmall by removing unused components, streamlining the code.

limichange committed:
refactor: Remove unused transaction components and streamline layout

- Deleted deprecated components: TransactionPrice, TransactionTime, and TransactionType to simplify the TransactionsHistory structure.
- Updated TransactionItemNormal and TransactionItemSmall to directly use SizableText for displaying transaction details, enhancing code clarity.
- Adjusted price formatting logic to conditionally display market cap or price based on value, improving responsiveness in the layout.

limichange committed:
refactor: Update TransactionsHeaderSmall layout and text identifiers

- Added flex property to TransactionsHeaderSmall for improved layout responsiveness.
- Swapped translation identifiers for time and type in the SizableText components to correct their display.
- Adjusted price width in useTransactionsLayoutSmall for better alignment.

limichange committed:
feat: Add left column width state management for responsive layout

- Introduced leftColumnWidthAtom to manage the width of the left column dynamically.
- Updated TransactionsHistory and DesktopLayout components to utilize the new atom for responsive design adjustments.
- Enhanced layout handling in DesktopLayout to capture and set the left column width on layout changes.

limichange committed:
fix: Update icon name in TransactionAddress component

- Changed the icon name from "Copy2Outline" to "Copy3Outline" in the TransactionAddress component for consistency with design specifications.

limichange committed:
refactor: Simplify TransactionsHistory components and enhance value formatting

- Removed unused imports in TransactionsHistory component for cleaner code.
- Updated TransactionItemNormal to use NumberSizeableText for value display, incorporating conditional formatting based on value size, improving clarity and responsiveness.

limichange committed:
fix: Set default liquidity filter value in MarketHome component

- Updated the liquidityFilter state in MarketHome to initialize with a minimum value of '5K', ensuring a more relevant default for users.

limichange committed:
feat: Add 'nodata' to spell checker skip words list

- Included 'nodata' in the spellCheckerSkipWords module to prevent false positives during spell checking.

limichange committed:
Merge branch 'x' into market-update-5

limichange committed:
refactor: Optimize imports and enhance type definitions in transaction components

- Removed unused imports in InformationTabs and updated TransactionAddress and TransactionAmount components to use specific type definitions for style props, improving type safety.
- Refactored value calculation in useTransactionItemData to utilize BigNumber for better precision in financial calculations.
- Cleaned up console log in DesktopLayout to streamline layout handling.

limichange committed:
feat: Enhance NumberSizeableText with autoFormatter functionality

- Introduced autoFormatter and autoFormatterThreshold props to NumberSizeableText for dynamic formatting based on value size.
- Updated TransactionItemNormal and TransactionItemSmall components to utilize the new autoFormatter feature for improved clarity in price and value display.
- Added examples in NumberSizeableTextGallery to demonstrate various formatting scenarios based on value thresholds.

limichange committed:
Merge branch 'x' into market-update-5

limichange committed:
feat: Enhance MarketTokenList with sorting functionality for watchlist and main list

- Added sortIndex property to IMarketToken interface for improved sorting capabilities.
- Refactored MarketTokenList to manage separate sorting states for the main list and watchlist.
- Updated useMarketWatchlistTokenList hook to utilize sortIndex for natural ordering in watchlist.
- Modified transformApiItemToToken function to include sortIndex during token transformation.
- Improved addIntoWatchListV2 action to calculate sortIndex for new items, ensuring they appear at the top of the watchlist.

limichange committed:
refactor: Simplify sorting logic in MarketTokenList and update hooks

- Removed separate sorting states for main list and watchlist in MarketTokenList.
- Consolidated sorting functionality into a single handleSortChange method.
- Updated useMarketTokenList and useMarketWatchlistTokenList hooks to accept initialSortBy and initialSortType parameters for improved flexibility.
- Enhanced state management for sorting within the hooks to streamline data handling.

limichange committed:
refactor: Update MarketTokenList integration in mobile and desktop layouts

- Replaced defaultShowWatchlistOnly prop with externalWatchlistControl in MarketMobileTabs for better control over watchlist visibility.
- Removed defaultShowWatchlistOnly from DesktopLayout, utilizing externalWatchlistControl instead to streamline the component's logic.

limichange committed:
refactor: Improve watchlist update mechanism in MarketV2 actions and hooks

- Modified removeFromWatchListV2 action to immediately update local state before making an asynchronous API call, enhancing user experience.
- Updated useMarketWatchlistTokenList hook to filter transformed data based on the current watchlist, ensuring immediate UI updates reflect changes.

limichange committed:
refactor: Streamline addIntoWatchListV2 action and enhance MarketTokenList sorting

- Simplified addIntoWatchListV2 action by removing the async keyword, allowing immediate local state updates before the API call.
- Updated MarketTokenList to ensure consistent sorting behavior by removing conditional sorting logic and enhancing the useEffect for better state management.
- Added a unique key to the MarketTokenList table for improved rendering performance.

limichange committed:
refactor: Simplify watchlist control in MarketTokenList and related components

- Replaced externalWatchlistControl prop with a direct showWatchlistOnly boolean in MarketTokenList for clearer control over watchlist visibility.
- Updated MarketMobileTabs and DesktopLayout to utilize the new showWatchlistOnly prop, streamlining the component logic.
- Enhanced useMarketWatchlistTokenList hook with additional revalidation options for improved data consistency.

limichange committed:
feat: Introduce MarketWatchListProviderV2 and integrate into MarketDetailV2 and MarketHomeV2

- Added MarketWatchListProviderV2 and MarketWatchListProviderMirrorV2 components to enhance watchlist functionality.
- Updated MarketDetailV2 and MarketHomeV2 to utilize the new providers for improved state management and context handling.
- Registered marketWatchListV2 in the context store map for better integration.

limichange committed:
feat: Implement showWatchlistOnly functionality in MarketV2

- Added showWatchlistOnly atom and actions to manage watchlist visibility state.
- Integrated showWatchlistOnly actions into MarketHomeV2, MarketMobileTabs, and MarketTokenList components for dynamic UI updates.
- Emitted app events on showWatchlistOnly changes to synchronize state across components.
- Updated MarketViewToggle to control watchlist visibility based on user interaction.

limichange committed:
fix: Ensure consistent displayName assignment in Market components

- Added missing newline at the end of the file for MarketWatchListProviderMirrorV2 and MarketWatchListProviderV2.
- Updated MarketDetailV2 and MarketHomeV2 to format the MarketWatchListProviderMirrorV2 component for better readability.

sidmorizon commented on packages/kit/src/states/jotai/contexts/marketV2/actions.ts:
```diff
@@ -87,35 +111,46 @@ class ContextJotaiActionsMarketV2 extends ContextJotaiActionsBase {
+      set(marketWatchListV2Atom(), { ...prev, data: newData });
+
+      // 异步调用 API，不等待结果
+      void backgroundApiProxy.serviceMarketV2.removeMarketWatchListV2({
```
假如 background 报错失败，是否需要还原数据

limichange replied:
给云端同步留的，先注释掉。

sidmorizon commented on packages/kit/src/states/jotai/contexts/marketV2/actions.ts:
```diff
@@ -228,3 +263,22 @@ export function useTokenDetailActions() {
+  const toggleShowWatchlistOnly = actions.toggleShowWatchlistOnly.use();
+
+  return useRef({
+    setShowWatchlistOnly: setShowWatchlistOnly as (value: boolean) => void,
```
这个类型是能推导的吧，不需要 as，其他地方的代码都没有 as

sidmorizon commented on packages/kit/src/views/Market/MarketDetailV2/components/InformationTabs/components/TransactionsHistory/hooks/useTransactionItemData.ts:
```diff
@@ -0,0 +1,73 @@
+  const intl = useIntl();
+  const { copyText } = useClipboard();
+
+  const handleCopyAddress = () => {
```
useCallback 和 useMemo 处理，否则容易导致上层调用的 hooks 重复执行

limichange committed:
refactor: Update comments for clarity and remove console logs in Market components

- Translated comments in actions.ts to English for better understanding.
- Removed unnecessary console log statements in MarketTokenList and watchListHooksV2 for cleaner code.

sidmorizon commented on packages/kit/src/views/Market/MarketDetailV2/components/TokenDetailHeader/TokenDetailHeaderRight.tsx:
```diff
@@ -95,9 +99,13 @@ export function TokenDetailHeaderRight({
-          numberFormat(marketCap, { formatter: 'marketCap' }),
-        )}`}
+        value={
+          formatDisplayNumber(
```
判断类型 string 正常处理，否则传 undefined ，直接 as 类型不对，底层组件渲染的时候可能导致 app 闪退

limichange committed:
Merge branch 'x' into market-update-5

sidmorizon commented on packages/kit/src/views/Market/MarketDetailV2/layouts/DesktopLayout.tsx:
```diff
@@ -9,8 +10,16 @@ import {
   const { tokenAddress, networkId, tokenDetail } = useTokenDetail();
+  const [, setLeftColumnWidth] = useLeftColumnWidthAtom();
+
+  const handleLayout = (event: LayoutChangeEvent) => {
```
useCallback

limichange committed:
refactor: Remove moveToTopV2 action and simplify showWatchlistOnly actions

- Deleted the moveToTopV2 action from ContextJotaiActionsMarketV2 to streamline functionality.
- Simplified the return structure of useShowWatchlistOnlyActions for improved clarity and type safety.

limichange committed:
refactor: Optimize handleLayout function in DesktopLayout component

- Wrapped handleLayout function in useCallback to improve performance and prevent unnecessary re-renders.
- Ensured setLeftColumnWidth dependency is included for better stability in the layout handling logic.

limichange committed:
refactor: Optimize address copy and time formatting in Market components

- Wrapped handleCopyAddress and handleViewInBrowser functions in useCallback to enhance performance and prevent unnecessary re-renders.
- Updated formatRelativeTime function to use useCallback for improved efficiency.
- Refactored value props in TokenDetailHeaderRight to utilize NumberSizeableText for better formatting and display consistency.

limichange committed:
refactor: Comment out asynchronous API calls in marketV2 actions

- Commented out the addMarketWatchListV2 and removeMarketWatchListV2 API calls in ContextJotaiActionsMarketV2 to prevent unintended execution during state updates.
- This change aims to enhance code clarity and maintainability while ensuring that the actions can be re-enabled when needed.

limichange committed:
Merge branch 'x' into market-update-5

limichange committed:
refactor: Enhance TradingView integration with locale and timezone support

- Replaced hardcoded locale mapping with a dedicated tradingViewLocaleMap utility for improved maintainability.
- Introduced getTradingViewTimezone utility to dynamically fetch the user's timezone.
- Updated TradingViewV2 component to utilize the new locale and timezone functionalities, ensuring accurate display settings in the TradingView URL.

limichange committed:
refactor: Simplify formatRelativeTime function in useTransactionItemData hook

- Removed useCallback wrapper from formatRelativeTime function for improved readability and simplicity.
- This change enhances code clarity while maintaining the same functionality.

limichange committed:
Merge branch 'x' into market-update-5

limichange committed:
refactor: Optimize typeText computation in useTransactionItemData hook

- Replaced direct conditional logic for typeText with useMemo to enhance performance and prevent unnecessary recalculations.
- This change improves the efficiency of the hook while maintaining the same functionality.

limichange committed:
refactor: Simplify locale handling in TradingViewV2 component

- Updated locale assignment in TradingViewV2 to directly use systemLocale, removing reliance on tradingViewLocaleMap for improved clarity and maintainability.
- This change streamlines the URL parameter setup while ensuring accurate locale representation.

limichange committed:
Merge branch 'x' into market-update-5
